src/GafferScene/BranchCreator.cpp

```diff
@@ -539,102 +724,141 @@ bool BranchCreator::processesRootObject() const
-	}
-	else if( destinationMatch == IECore::PathMatcher::ExactMatch && processesRootObject() )
+	switch( locationType )
 	{
-		// See notes in `hashObject()`.
-		if( !destinationPlug()->isSetToDefault() )
-		{
-			throw IECore::Exception( "Can only process root object when `destination` is default." );
-		}
-		hashBranchObject( path, branchPath, context, h );
-	}
-	else
-	{
-		h = inPlug()->objectPlug()->hash();
+		case Branch :
+			hashBranchObject( sourcePath, branchPath, context, h );
+			break;
+		case Destination :
+			if( processesRootObject() )
+			{
+				// See notes in `hashObject()`.
+				if( !destinationPlug()->isSetToDefault() )
+				{
+					throw IECore::Exception( "Can only process root object when `destination` is default." );
+				}
+				hashBranchObject( path, branchPath, context, h );
+				break;
+			}
+			// Fall through
+		case Ancestor :
+		case PassThrough :
+			h = inPlug()->objectPlug()->hash();
+			break;
+		default :
+			h = inPlug()->objectPlug()->defaultHash();
 	}
 }
 
 IECore::ConstObjectPtr BranchCreator::computeObject( const ScenePath &path, const Gaffer::Context *context, const ScenePlug *parent ) const
 {
 	ScenePath sourcePath, branchPath;
-	const IECore::PathMatcher::Result destinationMatch = sourceAndBranchPaths( path, sourcePath, branchPath );
+	const LocationType locationType = sourceAndBranchPaths( path, sourcePath, branchPath );
 
-	if( destinationMatch == IECore::PathMatcher::AncestorMatch )
-	{
-		return computeBranchObject( sourcePath, branchPath, context );
-	}
-	else if( destinationMatch == IECore::PathMatcher::ExactMatch && processesRootObject() )
+	switch( locationType )
 	{
-		/// \todo The `processesRootObject()` mechanism was intended to allow derived
-		/// classes to modify the object at `sourcePath`, but it is not compatible
-		/// with the new `destination` plug. We could continue to support it by doing
-		/// another filter evaluation, but the mechanism of processing the "root"
-		/// object feels bogus now that source and destination are decoupled. Come up
-		/// with something more logical.
-		if( !destinationPlug()->isSetToDefault() )
-		{
-			throw IECore::Exception( "Can only process root object when `destination` is default." );
-		}
-		return computeBranchObject( path, branchPath, context );
-	}
-	else
-	{
-		return inPlug()->objectPlug()->getValue();
+		case Branch :
+			return computeBranchObject( sourcePath, branchPath, context );
+		case Destination :
+			if( processesRootObject() )
+			{
+				/// \todo The `processesRootObject()` mechanism was intended to allow derived
+				/// classes to modify the object at `sourcePath`, but it is not compatible
+				/// with the new `destination` plug. We could continue to support it by doing
+				/// another filter evaluation, but the mechanism of processing the "root"
+				/// object feels bogus now that source and destination are decoupled. Come up
+				/// with something more logical.
```

Comment:
Do we have any real use for processing the root object other than deleting it in Instancer?  Seems like it might be valid to say that BranchCreators in general don't process the root object, and just add a DeleteObject to Instancer ( I guess the way this would get wired nicely up is a bit tricky, with BranchCreator::inPlug() being hidden and driven by the DeleteObject, with the Instancer able to access the visible input plug to read the objects before they get deleted ). 
 Not that this needs to happen immediately, but just thinking.

Reply:
I think we also use it in Unencapsulate, to replace the capsule with the object at the root of the capsule scene. I like your DeleteObject gambit, but I don't think that'll work for this use case...

Reply:
Hmm, interesting.  Unencapsulate does feel like a bit of a different case, where modifying the destination really doesn't make any sense ( whereas it could be useful in Instancer ), but drawing the line between these is kinda tricky.

Reply:
Yeah, I think never exposing `destination` on Unencapsulate is probably pretty reasonable, so maybe DeleteObject will be the way to go. I think the way to do the DeleteObject is probably not to hide `BranchCreator.in`, but to add `BranchCreator.__preprocessedIn` with a default connection from `in`, so that Instancer can insert its DeleteObject in the middle?

Reply:
Sounds pretty good.